### ISSUE TITLE
torch output vs neuron output --> different value

### ISSUE THREAD

**Issue by byeongcheoljo** [aws-neuron/aws-neuron-sdk#765]
i am using inf1 EC2.

I created a model using pytorch and converted it into a neuron model. When comparing the two outputs, there is a significant difference in the resulting values.

I would like to inquire about what part needs to be modified.

torch-->neuron model code
---------------------------
```
import torch 
import torch.neuron
from transformers import BertForSequenceClassification, BertJapaneseTokenizer
import numpy as np
import os 

class Compile_neuron_model():
    def __init__(self,input_text):
        self.input_text = input_text
        self.model_name = "model_name"
        self.model = BertForSequenceClassification.from_pretrained(self.model_name)
        self.tokenizer = BertJapaneseTokenizer.from_pretrained(self.model_name)
        
    def compile_model(self):
        inputs = self.tokenizer(self.input_text, max_length=512, padding="max_length", truncation=True, return_tensors="pt", return_token_type_ids=True)
        
        kwargs = {'compiler_args':['--fast-math', 'none']}

        model_neuron = torch.neuron.trace(self.model, example_inputs=(inputs['input_ids'],inputs['token_type_ids'],inputs['attention_mask']), dynamic_batch_size=True, strict=False, **kwargs, verbose=0)
        save_dir = ""
        os.makedirs(save_dir, exist_ok=True)
        model_neuron.save(save_dir+"[jp_sentence_neuron_compile_fast-math_none.pt]")
        
        with torch.no_grad():
            output = self.model(**inputs)
        return list(output)

text = [
        "おかしい夢小説の見すぎじゃないのお前のセコム可哀想"]
print(Compile_neuron_model(text).compile_model())
```
```

[ 2.0653, -2.5520],
        [-0.1735,  0.3542],
        [ 3.1728, -3.4995],
        [ 2.5098, -3.0954],
        [ 0.1922, -0.0455],
        [ 2.5808, -2.6816],
        [ 3.1459, -3.4638],
        [ 0.4024, -0.3765],
        [-0.2901,  0.3380],
        [ 1.9861, -2.4405],
        [-1.5397,  1.9075],
        [ 3.3199, -3.5437],
        [ 2.6218, -3.1925],
        [ 2.6642, -3.0554],
        [-0.7760,  1.1985]
```

------------------------------------------------------------------------------------------

load and check output value in neuron model

```

model = torch.neuron.DataParallel(
            torch.jit.load(model_name))
tokenizer = BertJapaneseTokenizer.from_pretrained(tokenizer_model)
embedings = tokenizer(
            list(request.input_text),
            max_length=512,
            padding='max_length',
            truncation=True,
            return_tensors='pt',
            return_token_type_ids=True)
        
        outputs = model(*(embedings.values()))
        print(model(*(embedings['input_ids'], embedings['token_type_ids'], embedings['attention_mask'])))
        print(outputs)

```

```

[-0.3506,  0.0101],
        [-0.3584, -0.0064],
        [-0.2988, -0.0962],
        [-0.3655,  0.0214],
        [-0.3656,  0.0054],
        [-0.3628,  0.0284],
        [-0.3318, -0.0086],
        [-0.3434, -0.0254],
        [-0.3804, -0.0171],
        [-0.3333, -0.0280],
        [-0.3573,  0.0098],
        [-0.3409, -0.0280],
        [ 0.1563, -0.6810],
        [-0.2355, -0.1612],
        [-0.2919, -0.0882]]
```

**Comment by aws-donkrets** [comment]
Hi [byeongcheoljo](https://github.com/byeongcheoljo), 
Can you paste output from the following commands so report what version of the inf1 Neuron software you are using?
            $ yum list installed | grep neuron    ## on AL2
            $ apt list | grep neuron | grep installed   ## on Ubuntu

            $ pip3 list | grep neuron   ## both AL2 and Ubuntu
            $ cat /sys/devices/virtual/dmi/id/product_name    ## both AL2 and Ubuntu

**Comment by byeongcheoljo** [comment]
@aws-donkrets 

Hi, aws-donkrets.

The inf1 Neuron software versions are as follows:

`pip3 list | grep neuron`

```
neuron-cc                1.9.1.0+5f09e5c8e
torch-neuron             1.13.1.2.9.1.0
```

`cat /sys/devices/virtual/dmi/id/product_name`
```
inf1.xlarge
```

**Comment by sydsim** [comment]
Same here!!!!!!!!!
I also found that output changes depends on compiler arguments.
I tried with compiler_args=['--fast-math', 'fp32-cast-matmult'] and without any compiler_args, two outputs are different to each other and none of them are same with the correct output (made with nvidia gpu)

This is my environment
```
aws-neuronx-collectives.x86_64        2.17.9.0_fb6d14044-1           @neuron    
aws-neuronx-dkms.noarch               2.13.4.0-dkms                  @neuron    
aws-neuronx-oci-hook.x86_64           2.2.25.0-1                     @neuron    
aws-neuronx-runtime-lib.x86_64        2.17.7.0_df62e3f70-1           @neuron    
aws-neuronx-tools.x86_64              2.14.6.0-1                     @neuron    

neuron-cc                 1.20.3.0+ed6db4a2e
torch-neuron              1.13.1.2.9.6.0

inf1.xlarge
```
I wasted a couple of days because of this....

**Comment by aws-donkrets** [comment]
Hi [sydsim](https://github.com/sydsim)- Compiler arguments change the compiler's behavior and the code that it generates, so it is expected to get different results with different compiler arguments.  To help with the triage, can you try re-compiling and re-running your model with the `--fast-math none` flag to see if that improves the accuracy of the result.

Also, [byeongcheoljo](https://github.com/byeongcheoljo) - while I don't expect the results to change, you should consider downloading the latest Neuron SDK (which includes the neuron-cc 1.20.3.0+ed6db4a2e compiler version). This is the version we are using internally to triage your issue.

**Comment by hannanjgaws** [comment]
To add to what aws-donkrets said: On inf1 hardware, matmults are either performed in `bf16` or `fp16`. `fp32` matmults are not supported. One thing that may improve accuracy is to use `fp16` matmults rather than `bf16`. The most accurate `fp16`-optimized configuration can be enabled with:
```
--fast-math fp32-cast-matmult-fp16 no-fast-relayout
```

Also @byeongcheoljo or @sydsim are you able to point us to a checkpoint and tokenizer that reproduces the mismatch you're seeing? This will help us narrow down the problem and focus our debug efforts.

**Comment by sydsim** [comment]
@aws-donkrets 
I knew there would be a little difference, but I didn't know the difference would be this big to affect performance. 

@hannanjgaws 
I tried the arguments you suggested (compiler_args=['--fast-math', 'fp32-cast-matmult-fp16', 'no-fast-relayout']) and turned out there was a significant improvement to the accuracy, but still the difference is not neglectable. 
Different to byeongcheoljo's case, I'm not using open model like BERT but using private model, so it is not possible to share you the checkpoint. 

restructing the model to reduce the accuracy drop when using fp16 seems the only solution to use inf1 in my case now.

**Comment by aws-donkrets** [comment]
Just reviewing some older tickets. This one has been open for ~18 months. The newer SDK, along with the suggested compiler options, seem to have improved the situation so I'm going to close this out. If you need further assistance, please open a new ticket with results from the latest SDK. Thx.
